internal/receiver/discoveryreceiver/endpoint_tracker.go

```diff
@@ -214,40 +227,56 @@ func (n *notify) OnChange(changed []observer.Endpoint) {
 func entityStateEvents(observerID component.ID, endpoints []observer.Endpoint, correlations correlationStore, ts time.Time) (ees experimentalmetricmetadata.EntityEventsSlice, failed int, err error) {
 	entityEvents := experimentalmetricmetadata.NewEntityEventsSlice()
 	for _, endpoint := range endpoints {
+		if endpoint.Details == nil {
+			failed++
+			err = multierr.Combine(err, fmt.Errorf("endpoint %q has no details", endpoint.ID))
+			continue
+		}
+
 		entityEvent := entityEvents.AppendEmpty()
 		entityEvent.SetTimestamp(pcommon.NewTimestampFromTime(ts))
-		entityEvent.ID().PutStr(discovery.EndpointIDAttr, string(endpoint.ID))
 		entityState := entityEvent.SetEntityState()
 		attrs := entityState.Attributes()
-		if endpoint.Details != nil {
-			if envAttrs, e := endpointEnvToAttrs(endpoint.Details.Type(), endpoint.Details.Env()); e != nil {
-				err = multierr.Combine(err, fmt.Errorf("failed determining attributes for %q: %w", endpoint.ID, e))
-				failed++
-			} else {
-				// this must be the first mutation of attrs since it's destructive
-				envAttrs.CopyTo(attrs)
-			}
-			attrs.PutStr("type", string(endpoint.Details.Type()))
+		if envAttrs, e := endpointEnvToAttrs(endpoint.Details.Type(), endpoint.Details.Env()); e != nil {
+			err = multierr.Combine(err, fmt.Errorf("failed determining attributes for %q: %w", endpoint.ID, e))
+			failed++
+		} else {
+			// this must be the first mutation of attrs since it's destructive
+			envAttrs.CopyTo(attrs)
 		}
+		attrs.PutStr("type", string(endpoint.Details.Type()))
+		attrs.PutStr(discovery.EndpointIDAttr, string(endpoint.ID))
 		attrs.PutStr("endpoint", endpoint.Target)
 		attrs.PutStr(observerNameAttr, observerID.Name())
 		attrs.PutStr(observerTypeAttr, observerID.Type().String())
 		for k, v := range correlations.Attrs(endpoint.ID) {
 			attrs.PutStr(k, v)
 		}
+		extractIdentifyingAttrs(attrs, entityEvent.ID())
 	}
 	return entityEvents, failed, err
 }
 
-func entityDeleteEvents(endpoints []observer.Endpoint, ts time.Time) experimentalmetricmetadata.EntityEventsSlice {
+func entityDeleteEvents(endpoints []observer.Endpoint, ts time.Time) (ees experimentalmetricmetadata.EntityEventsSlice, failed int, err error) {
```

Comment:
Why we need the failed (int) value? Is it a return code?

Reply:
This is just for logging. We do the same for `emitEntityStateEvents`, so I kept it the same, but I agree it's not ideal. We can probably move the logging inside the funcs. I can address it separately